### DISCUSSION TITLE
What message delivery / op processing guarantees does Dagster have?

### DISCUSSION THREAD

**Discussion by skeller88** [dagster-io/dagster#20113] [Q&A]
In a distributed system, guarantees are hard. In systems with queues like Kafka or RabbitMQ, there's usually a guarantee of at least once, at most once, or exactly once processing: https://blog.bytebytego.com/p/at-most-once-at-least-once-exactly. What are Dagster's guarantees or guidelines around that?

**Comment by yuhan** [comment]
Curious what exactly you meant by the message guarantees in Dagster context?

Here are some relevant notes:
- Every schedule or sensor tick executes exactly once per tick.
- Within a pipeline run, every op or asset executes exactly once.
- In case you are looking for idempotency in kicking off runs, here's our guideline for sensor (which kicks off runs): https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#idempotence-using-run-keys

**Reply by skeller88** [reply]
That's amazing. I think I'm just surprised at an exactly once guarantee, because that's really hard to do in a distributed system. For example, say that you have an asset that generates a UUID and creates a s3 file with that UUID as the object key. The asset is materialized, and the s3 object is written successfully, but Dagster thinks that materialization fails for some reason, such as the run result failing to write to run storage. Then Dagster would attempt to rerun the asset materialization, and that would result in two s3 objects instead of the expected one. We
